File: android/app/src/main/java/com/clipsync/android/media/MediaBlobStore.kt
package com.clipsync.android.media

import java.io.File
import java.io.FileOutputStream
import java.security.MessageDigest
import java.util.UUID

class PendingMediaWrite internal constructor(
    internal val tempPath: File,
    internal val stream: FileOutputStream,
    internal val hasher: MessageDigest,
) {
    var bytesWritten: Long = 0
        internal set

    fun close() {
        runCatching { stream.close() }
    }
}

/**
 * Content-addressed PNG/JPEG blob store. Callers never concatenate blob paths;
 * every file name is a validated SHA-256 hex digest or a UUID temp name.
 */
class MediaBlobStore(rootDirectory: File) {
    private val root = rootDirectory.canonicalFile
    private val blobs = File(root, BLOBS_DIRECTORY)
    private val temps = File(root, TEMP_DIRECTORY)
    private val thumbs = File(root, THUMBS_DIRECTORY)

    init {
        blobs.mkdirs()
        temps.mkdirs()
        thumbs.mkdirs()
    }

    val rootDirectory: File get() = root

    fun beginWrite(): PendingMediaWrite {
        val tempPath = File(temps, UUID.randomUUID().toString().replace("-", "") + ".part")
        val stream = FileOutputStream(tempPath)
        return PendingMediaWrite(tempPath, stream, MessageDigest.getInstance("SHA-256"))
    }

    fun append(pending: PendingMediaWrite, chunk: ByteArray) {
        require(chunk.isNotEmpty()) { "Image chunks cannot be empty." }
        val next = pending.bytesWritten + chunk.size
        if (next > MediaLimits.MAX_ENCODED_BYTES) {
            throw MediaStoreException(MEDIA_TOO_LARGE, "Encoded image exceeds 16 MiB.")
        }
        pending.stream.write(chunk)
        pending.hasher.update(chunk)
        pending.bytesWritten = next
    }

    fun commit(
        pending: PendingMediaWrite,
        expectedHash: String? = null,
        expectedMime: String? = null,
    ): ValidatedImage {
        pending.stream.flush()
        val hash = pending.hasher.digest().joinToString("") { "%02x".format(it) }
        pending.close()
        try {
            if (expectedHash != null && expectedHash != hash) {
                throw MediaStoreException(MEDIA_HASH_MISMATCH, MEDIA_HASH_MISMATCH)
            }
            val (inspect, image) = ImageCodec.tryInspectFile(
                pending.tempPath,
                hash,
                pending.bytesWritten,
            )
            if (inspect != ImageCodecError.OK || image == null) {
                throw MediaStoreException(mapInspect(inspect), mapInspect(inspect))
            }
            if (expectedMime != null && expectedMime != image.mimeType) {
                throw MediaStoreException(UNSUPPORTED_MEDIA, UNSUPPORTED_MEDIA)
            }
            val destination = blobPath(image.contentHash)
            destination.parentFile?.mkdirs()
            if (destination.exists()) {
                pending.tempPath.delete()
                return image
            }
            if (!pending.tempPath.renameTo(destination)) {
                pending.tempPath.copyTo(destination, overwrite = false)
                pending.tempPath.delete()
            }
            return image
        } catch (error: Throwable) {
            pending.tempPath.delete()
            throw error
        }
    }

    fun commitBytes(encoded: ByteArray, expectedHash: String? = null): ValidatedImage {
        val (inspect, image) = ImageCodec.tryInspect(encoded, expectedHash)
        if (inspect != ImageCodecError.OK || image == null) {
            throw MediaStoreException(mapInspect(inspect), mapInspect(inspect))
        }
        val destination = blobPath(image.contentHash)
        if (destination.exists()) {
            return image
        }
        val pending = beginWrite()
        return try {
            append(pending, encoded)
            commit(pending, image.contentHash, image.mimeType)
        } catch (error: Throwable) {
            pending.close()
            pending.tempPath.delete()
            throw error
        }
    }

    fun exists(contentHash: String): Boolean = blobPath(contentHash).isFile

    fun requirePath(contentHash: String): File {
        val path = blobPath(contentHash)
        if (!path.isFile) {
            throw MediaStoreException(MEDIA_STORAGE_FAILED, "Media blob is missing.")
        }
        return path
    }

    fun readAllBytes(contentHash: String): ByteArray = requirePath(contentHash).readBytes()

    fun thumbnailPath(contentHash: String): File =
        File(thumbs, normalizeHash(contentHash) + ".png")

    fun deleteBlob(contentHash: String) {
        blobPath(contentHash).delete()
        thumbnailPath(contentHash).delete()
    }

    fun recoverTemps(nowMs: Long, maximumDeletes: Int = 256): Int {
        val cutoff = nowMs - MediaLimits.UNFINISHED_DOWNLOAD_HOURS * 60L * 60L * 1000L
        if (!temps.isDirectory) {
            return 0
        }
        var removed = 0
        temps.listFiles { file -> file.isFile && file.name.endsWith(".part") }?.forEach { file ->
            if (removed >= maximumDeletes) {
                return removed
            }
            if (file.lastModified() <= cutoff) {
                file.delete()
                removed++
            }
        }
        return removed
    }

    fun deleteUnreferenced(liveHashes: Collection<String>, maximumDeletes: Int = 256): Int {
        val live = liveHashes.toHashSet()
        if (!blobs.isDirectory) {
            return 0
        }
        var removed = 0
        blobs.walkTopDown().filter { it.isFile }.forEach { file ->
            if (removed >= maximumDeletes) {
                return removed
            }
            val name = file.name
            if (name.length != 64 || name in live) {
                return@forEach
            }
            file.delete()
            File(thumbs, "$name.png").delete()
            removed++
        }
        return removed
    }

    fun blobPath(contentHash: String): File = File(blobs, normalizeHash(contentHash))

    companion object {
        const val BLOBS_DIRECTORY = "blobs"
        const val TEMP_DIRECTORY = "tmp"
        const val THUMBS_DIRECTORY = "thumbs"

        const val UNSUPPORTED_MEDIA = "UNSUPPORTED_MEDIA"
        const val MEDIA_TOO_LARGE = "MEDIA_TOO_LARGE"
        const val MEDIA_DECODE_FAILED = "MEDIA_DECODE_FAILED"
        const val MEDIA_HASH_MISMATCH = "MEDIA_HASH_MISMATCH"
        const val MEDIA_STORAGE_FAILED = "MEDIA_STORAGE_FAILED"

        fun defaultRootForDatabase(databaseFile: File): File =
            File(databaseFile.parentFile ?: File("."), "media")

        private fun normalizeHash(contentHash: String): String {
            require(contentHash.length == 64 && contentHash.all { it in '0'..'9' || it in 'a'..'f' }) {
                "content_hash must be 64 lowercase hex characters."
            }
            return contentHash
        }

        private fun mapInspect(error: ImageCodecError): String = when (error) {
            ImageCodecError.TOO_LARGE -> MEDIA_TOO_LARGE
            ImageCodecError.HASH_MISMATCH -> MEDIA_HASH_MISMATCH
            ImageCodecError.UNSUPPORTED_MEDIA -> UNSUPPORTED_MEDIA
            else -> MEDIA_DECODE_FAILED
        }
    }
}

class MediaStoreException(val code: String, message: String) : Exception(message)

File: windows/ClipSync.Core/Media/MediaBlobStore.cs
namespace ClipSync.Core.Media;

public sealed class PendingMediaWrite : IDisposable
{
    internal PendingMediaWrite(string tempPath, FileStream stream)
    {
        TempPath = tempPath;
        Stream = stream;
        Hasher = System.Security.Cryptography.IncrementalHash.CreateHash(
            System.Security.Cryptography.HashAlgorithmName.SHA256);
    }

    internal string TempPath { get; }

    internal FileStream Stream { get; }

    internal System.Security.Cryptography.IncrementalHash Hasher { get; }

    public long BytesWritten { get; internal set; }

    public void Dispose()
    {
        Stream.Dispose();
        Hasher.Dispose();
    }
}

/// <summary>
/// Content-addressed PNG/JPEG blob store. Callers never concatenate blob paths;
/// every file name is a validated SHA-256 hex digest or a UUID temp name.
/// </summary>
public sealed class MediaBlobStore
{
    public const string BlobsDirectoryName = "blobs";
    public const string TempDirectoryName = "tmp";
    public const string ThumbnailsDirectoryName = "thumbs";

    private readonly string root;
    private readonly string blobs;
    private readonly string temps;
    private readonly string thumbs;

    public MediaBlobStore(string rootDirectory)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(rootDirectory);
        root = Path.GetFullPath(rootDirectory);
        blobs = Path.Combine(root, BlobsDirectoryName);
        temps = Path.Combine(root, TempDirectoryName);
        thumbs = Path.Combine(root, ThumbnailsDirectoryName);
        Directory.CreateDirectory(blobs);
        Directory.CreateDirectory(temps);
        Directory.CreateDirectory(thumbs);
    }

    public string RootDirectory => root;

    public static string DefaultRootForDatabase(string databasePath)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(databasePath);
        var directory = Path.GetDirectoryName(Path.GetFullPath(databasePath));
        return Path.Combine(string.IsNullOrEmpty(directory) ? "." : directory, "media");
    }

    public PendingMediaWrite BeginWrite()
    {
        var tempPath = Path.Combine(temps, Guid.NewGuid().ToString("N") + ".part");
        var stream = new FileStream(
            tempPath,
            FileMode.CreateNew,
            FileAccess.Write,
            FileShare.None,
            bufferSize: 64 * 1024,
            FileOptions.SequentialScan);
        return new PendingMediaWrite(tempPath, stream);
    }

    public static void Append(PendingMediaWrite pending, ReadOnlySpan<byte> chunk)
    {
        ArgumentNullException.ThrowIfNull(pending);
        if (chunk.Length == 0)
        {
            throw new InvalidDataException("Image chunks cannot be empty.");
        }

        var next = pending.BytesWritten + chunk.Length;
        if (next > MediaLimits.MaxEncodedBytes)
        {
            throw new InvalidDataException("Encoded image exceeds 16 MiB.");
        }

        pending.Stream.Write(chunk);
        pending.Hasher.AppendData(chunk);
        pending.BytesWritten = next;
    }

    public ValidatedImage Commit(PendingMediaWrite pending, string? expectedHash = null, string? expectedMime = null)
    {
        ArgumentNullException.ThrowIfNull(pending);
        pending.Stream.Flush();
        var hash = Convert.ToHexString(pending.Hasher.GetHashAndReset()).ToLowerInvariant();
        pending.Dispose();

        try
        {
            if (expectedHash is not null && !string.Equals(expectedHash, hash, StringComparison.Ordinal))
            {
                throw new InvalidDataException("MEDIA_HASH_MISMATCH");
            }

            var inspect = ImageCodec.TryInspectFile(pending.TempPath, out var image, hash, pending.BytesWritten);
            if (inspect != ImageCodecError.Ok || image is null)
            {
                throw new InvalidDataException(inspect switch
                {
                    ImageCodecError.TooLarge => "MEDIA_TOO_LARGE",
                    ImageCodecError.HashMismatch => "MEDIA_HASH_MISMATCH",
                    ImageCodecError.UnsupportedMedia => "UNSUPPORTED_MEDIA",
                    _ => "MEDIA_DECODE_FAILED"
                });
            }

            if (expectedMime is not null && !string.Equals(expectedMime, image.MimeType, StringComparison.Ordinal))
            {
                throw new InvalidDataException("UNSUPPORTED_MEDIA");
            }

            var destination = BlobPath(image.ContentHash);
            Directory.CreateDirectory(Path.GetDirectoryName(destination)!);
            if (File.Exists(destination))
            {
                File.Delete(pending.TempPath);
                return image;
            }

            File.Move(pending.TempPath, destination);
            return image;
        }
        catch
        {
            TryDelete(pending.TempPath);
            throw;
        }
    }

    public ValidatedImage CommitBytes(ReadOnlySpan<byte> encoded, string? expectedHash = null)
    {
        var inspect = ImageCodec.TryInspect(encoded, out var image, expectedHash);
        if (inspect != ImageCodecError.Ok || image is null)
        {
            throw new InvalidDataException(inspect switch
            {
                ImageCodecError.TooLarge => "MEDIA_TOO_LARGE",
                ImageCodecError.HashMismatch => "MEDIA_HASH_MISMATCH",
                ImageCodecError.UnsupportedMedia => "UNSUPPORTED_MEDIA",
                _ => "MEDIA_DECODE_FAILED"
            });
        }

        var destination = BlobPath(image.ContentHash);
        if (File.Exists(destination))
        {
            return image;
        }

        var pending = BeginWrite();
        try
        {
            Append(pending, encoded);
            return Commit(pending, image.ContentHash, image.MimeType);
        }
        catch
        {
            pending.Dispose();
            TryDelete(pending.TempPath);
            throw;
        }
    }

    public bool Exists(string contentHash) => File.Exists(BlobPath(contentHash));

    public string RequirePath(string contentHash)
    {
        var path = BlobPath(contentHash);
        if (!File.Exists(path))
        {
            throw new FileNotFoundException("Media blob is missing.");
        }

        return path;
    }

    public Stream OpenRead(string contentHash) =>
        new FileStream(RequirePath(contentHash), FileMode.Open, FileAccess.Read, FileShare.Read);

    public byte[] ReadAllBytes(string contentHash) => File.ReadAllBytes(RequirePath(contentHash));

    public string ThumbnailPath(string contentHash) =>
        Path.Combine(thumbs, NormalizeHash(contentHash) + ".png");

    public void DeleteBlob(string contentHash)
    {
        TryDelete(BlobPath(contentHash));
        TryDelete(ThumbnailPath(contentHash));
    }

    /// <summary>Deletes temp files older than 24 hours. Caps the number of deletions per call.</summary>
    public int RecoverTemps(DateTimeOffset now, int maximumDeletes = 256)
    {
        var cutoff = now.AddHours(-MediaLimits.UnfinishedDownloadHours);
        var removed = 0;
        if (!Directory.Exists(temps))
        {
            return 0;
        }

        foreach (var file in Directory.EnumerateFiles(temps, "*.part"))
        {
            if (removed >= maximumDeletes)
            {
                break;
            }

            var info = new FileInfo(file);
            if (info.LastWriteTimeUtc <= cutoff.UtcDateTime)
            {
                TryDelete(file);
                removed++;
            }
        }

        return removed;
    }

    public int DeleteUnreferenced(IReadOnlyCollection<string> liveHashes, int maximumDeletes = 256)
    {
        ArgumentNullException.ThrowIfNull(liveHashes);
        var live = new HashSet<string>(liveHashes, StringComparer.Ordinal);
        var removed = 0;
        if (!Directory.Exists(blobs))
        {
            return 0;
        }

        foreach (var file in Directory.EnumerateFiles(blobs, "*", SearchOption.AllDirectories))
        {
            if (removed >= maximumDeletes)
            {
                break;
            }

            var name = Path.GetFileName(file);
            if (name.Length != 64 || live.Contains(name))
            {
                continue;
            }

            TryDelete(file);
            TryDelete(Path.Combine(thumbs, name + ".png"));
            removed++;
        }

        return removed;
    }

    internal string BlobPath(string contentHash)
    {
        var hash = NormalizeHash(contentHash);
        return Path.Combine(blobs, hash);
    }

    private static string NormalizeHash(string contentHash)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(contentHash);
        if (contentHash.Length != 64)
        {
            throw new ArgumentException("content_hash must be 64 lowercase hex characters.", nameof(contentHash));
        }

        foreach (var character in contentHash)
        {
            if (character is not ((>= '0' and <= '9') or (>= 'a' and <= 'f')))
            {
                throw new ArgumentException("content_hash must be 64 lowercase hex characters.", nameof(contentHash));
            }
        }

        return contentHash;
    }

    private static void TryDelete(string path)
    {
        try
        {
            if (File.Exists(path))
            {
                File.Delete(path);
            }
        }
        catch (IOException)
        {
        }
        catch (UnauthorizedAccessException)
        {
        }
    }
}
